File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithFactories12.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

@KoraApp
public interface AppWithFactories12 {
    @Root
    default Object mock1(ConfigValueExtractor<TestEnum> object) {
        return new Object();
    }

    default <T extends Enum<T>> EnumConfigValueExtractor<T> enumConfigValueExtractor() {
        return new EnumConfigValueExtractor<>();
    }

    enum TestEnum {}

    interface ConfigValueExtractor<T> {
        T extract(Object value);
    }

    class EnumConfigValueExtractor<T extends Enum<T>> implements ConfigValueExtractor<T> {
        @Override
        public T extract(Object value) {
            return null;
        }
    }

}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithFactories12.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root

@KoraApp
interface AppWithFactories12 {

    fun <T, R> factory1(t: SimpleGeneric<T>, r: SimpleGeneric<R>): GenericClass<T, R> {
        return GenericClass(t, r)
    }

    fun factory2(): SimpleGeneric<Long> {
        return SimpleGeneric()
    }

    fun factory3(): SimpleGeneric<String> {
        return SimpleGeneric()
    }

    @Root
    fun mock1(cl: GenericClass<Long, String>): Any = Any()

    open class SimpleGeneric<T>
    data class GenericClass<T, R>(val t: SimpleGeneric<T>, val r: SimpleGeneric<R>)
}
